cli/diag.c

```diff
@@ -110,6 +110,296 @@ static int diag_parse_common_cfg(int argc, char **argv, const char *desc,
+		{"ltssm", 'l', "", CFG_NONE, &cfg.enable_ltssm, no_argument,
+		 "Enable LTSSMloopback mode"},
+		{"rx-to-tx", 'r', "", CFG_NONE, &cfg.enable_rx_to_tx, no_argument,
+		 "Enable RX->TX loopback mode"},
+		{"tx-to-rx", 't', "", CFG_NONE, &cfg.enable_tx_to_rx, no_argument,
+		 "Enable TX->RX loopback mode"},
+		{"speed", 's', "GEN", CFG_CHOICES, &cfg.speed, required_argument,
+		 "LTSSM Speed (if enabling the LTSSM loopback mode), default: GEN4",
+		 .choices = loopback_ltssm_speeds},
+		{NULL}};
+
+	argconfig_parse(argc, argv, CMD_DESC_LOOPBACK, opts, &cfg, sizeof(cfg));
+
+	if (cfg.port_id < 0) {
+		fprintf(stderr, "Must specify -p / --port_id\n");
+		return -1;
+	}
+
+	if (cfg.disable && (cfg.enable_rx_to_tx || cfg.enable_tx_to_rx ||
+			    cfg.enable_ltssm)) {
+		fprintf(stderr,
+			"Must not specify -d / --disable with an enable flag\n");
+		return -1;
+	}
+
+	ret = get_port(cfg.dev, cfg.port_id, &cfg.port);
+	if (ret)
+		return ret;
+
+	if (cfg.disable || cfg.enable_rx_to_tx || cfg.enable_tx_to_rx ||
+	    cfg.enable_ltssm) {
+		if (cfg.enable_rx_to_tx)
+			enable |= SWITCHTEC_DIAG_LOOPBACK_RX_TO_TX;
+		if (cfg.enable_tx_to_rx)
+			enable |= SWITCHTEC_DIAG_LOOPBACK_TX_TO_RX;
+		if (cfg.enable_ltssm)
+			enable |= SWITCHTEC_DIAG_LOOPBACK_LTSSM;
+
+		ret = switchtec_diag_loopback_set(cfg.dev, cfg.port_id,
+						  enable, cfg.speed);
+		if (ret) {
+			switchtec_perror("loopback_set");
+			return -1;
+		}
+	}
+
+	return print_loopback_mode(cfg.dev, cfg.port_id);
+}
+
+static const struct argconfig_choice pattern_types[] = {
+	{"PRBS7",   SWITCHTEC_DIAG_PATTERN_PRBS_7,  "PRBS 7"},
+	{"PRBS11",  SWITCHTEC_DIAG_PATTERN_PRBS_11, "PRBS 11"},
+	{"PRBS23",  SWITCHTEC_DIAG_PATTERN_PRBS_23, "PRBS 23"},
+	{"PRBS31",  SWITCHTEC_DIAG_PATTERN_PRBS_31, "PRBS 31"},
+	{"PRBS9",   SWITCHTEC_DIAG_PATTERN_PRBS_9,  "PRBS 9"},
+	{"PRBS15",  SWITCHTEC_DIAG_PATTERN_PRBS_15, "PRBS 15"},
+	{}
+};
+
+static const char *pattern_to_str(enum switchtec_diag_pattern type)
+{
+	const struct argconfig_choice *s;
+
+	for (s = pattern_types; s->name; s++) {
+		if (s->value == type)
+			return s->name;
+	}
+
+	return "UNKNOWN";
+}
+
+static int print_pattern_mode(struct switchtec_dev *dev,
+		struct switchtec_status *port, int port_id)
+{
+	enum switchtec_diag_pattern gen_pat, mon_pat;
+	unsigned long long err_cnt;
+	int ret, lane_id;
+
+	ret = switchtec_diag_pattern_gen_get(dev, port_id, &gen_pat);
+	if (ret) {
+		switchtec_perror("pattern_gen_get");
+		return -1;
+	}
+
+	ret = switchtec_diag_pattern_mon_get(dev, port_id, 0, &mon_pat, &err_cnt);
+	if (ret) {
+		switchtec_perror("pattern_mon_get");
+		return -1;
+	}
+
+	printf("Port: %d\n", port_id);
+	if (gen_pat == SWITCHTEC_DIAG_PATTERN_PRBS_DISABLED)
+		printf("  Generator: Disabled\n");
+	else
+		printf("  Generator: %s\n", pattern_to_str(gen_pat));
+
+	if (mon_pat == SWITCHTEC_DIAG_PATTERN_PRBS_DISABLED) {
+		printf("  Monitor: Disabled\n");
+	} else {
+		printf("  Monitor: %-20s %d\n", pattern_to_str(mon_pat), mon_pat);
```

Comment:
Do we need the 'mon_pat' printed given we already have the string of it printed?

Reply:
Oh, that was a debug print... Removed